securedrop_client/api_jobs/base.py

```diff
@@ -30,29 +30,35 @@ class ApiJob(QObject):
     '''
     failure_signal = pyqtSignal(Exception)
 
-    def __init__(self) -> None:
+    def __init__(self, remaining_attempts: int = 5) -> None:
         super().__init__(None)  # `None` because the QOjbect has no parent
+        self.remaining_attempts = remaining_attempts
 
     def _do_call_api(self, api_client: API, session: Session) -> None:
         if not api_client:
             raise ApiInaccessibleError()
 
-        try:
-            result = self.call_api(api_client, session)
-        except AuthError as e:
-            raise ApiInaccessibleError() from e
-        except ApiInaccessibleError as e:
-            # Do not let ApiInaccessibleError emit the regular failure signal, raise now
-            raise ApiInaccessibleError() from e
-        except RequestTimeoutError:
-            logger.debug('Job {} timed out'.format(self))
-            raise
-        except Exception as e:
-            logger.error('Job {} raised an exception: {}: {}'
-                         .format(self, type(e).__name__, e))
-            self.failure_signal.emit(e)
-        else:
-            self.success_signal.emit(result)
+        while self.remaining_attempts >= 1:
+            try:
+                self.remaining_attempts -= 1
+                result = self.call_api(api_client, session)
+            except AuthError as e:
+                raise ApiInaccessibleError() from e
+            except ApiInaccessibleError as e:
+                # Do not let ApiInaccessibleError emit the regular failure signal, raise now
+                raise ApiInaccessibleError() from e
```

Comment:
Okay, so retry state is kept within a Job object, and the queue will just keep retrying that job if it gets either `ApiInaccessibleError` or `RequestTimeoutError`. So it's up to the job to immediately return after a certain number of tries.

Reply:
`AuthError` and `ApiInaccessibleError` both should raise (and upon raising those exceptions, execution should break out of the while loop) which will get handled in the queue and result in only a log line at this stage: https://github.com/freedomofpress/securedrop-client/blob/master/securedrop_client/queue.py#L47. The idea here is that if an `AuthError` or `ApiInaccessibleError` has occurred, there's no point in retrying the job as it means that e.g. the token expired or the user has logged out. 

At some later date we should improve the behavior here, e.g. implementing the suggestions in https://github.com/freedomofpress/securedrop-client/issues/379#issuecomment-501474662 (these probably need to get done in two different PRs given the scope of each change).

Reply:
Ah I thought the queue was catching these exceptions and logging and continuing to retry the job the way it does with RequestTimeoutErrors.

Reply:
The back and forth between catching and throwing throws me off a little. But I see that it doesn't set `last_job` now so the queue won't try again.